Add a RecogMethod type

ssiccha opened the pull request:
Fixes #258.

Things to double-check before this gets merged:
- did I remove all the hacks?
- did I update the method HOWTO?

ssiccha commented:
When I'm done I must check whether I forgot to remove any hacks.

ssiccha commented:
I've got a first version of `RecogMethod`s working. I've introduced several hacks so that I'm able to incrementally change every method into a `RecogMethod` object. When I'm finished I'm going to remove them again.

ssiccha commented:
I like what I've done so far. However, before I start converting all methods to `RecogMethod`s, could you check whether you disagree with some of my design choices @fingolfin? I think you can ignore `methsel.gi` and `recognition.gi` for now since they contain mostly hacks.

fingolfin commented on gap/base/methods.gd:
```diff
@@ -0,0 +1,40 @@
+#! method     : the function itself
+#! rank       : an integer rank
+#! stamp      : a string describing the method uniquely
+#! comment    : an optional comment to describe the method for humans
```
I assume the above comment is a draft and perhaps not meant for `IsRecogMethod` but rather for `RecogMethod`?

fingolfin commented on gap/base/methods.gi:
```diff
@@ -0,0 +1,59 @@
+function(stamp, comment, func)
+    local r;
+    r := rec();
+    Objectify(RecogMethodType, r);
```
Perhaps use `ObjectifyWithAttributes`, it can have better performance

fingolfin commented on gap/base/methods.gd:
```diff
@@ -0,0 +1,40 @@
+# Functions and attributes for IsRecogMethod objects
+DeclareGlobalFunction("RecogMethod");
+DeclareGlobalFunction("CallRecogMethod");
+DeclareAttribute("FunctionOfRecogMethod", IsRecogMethod);
+DeclareAttribute("Stamp", IsRecogMethod);
+DeclareAttribute("Comment", IsRecogMethod);
```
It's a bit odd that one has the `...OfRecogMethod` suffix but the others have not... I realize `Function` isn't a great name, though :-).

But anyway, why do we even need `FunctionOfRecogMethod `? Only `CallRecogMethod` needs it, right? So that could just access the value directly (you can store things into objects in other ways than using an attribute!).

ssiccha replied:
The purpose of the `FunctionOfRecogMethod` is to provide a documented access to the underlying function. I think people should have acces to that and I didn't want to document access the function via e.g. `m!.func`.
But I think instead of the attribute I'll store the function in `m!.func` and provide an access function `Unpack`.

ssiccha replied:
I've added a global function `UnpackRecogMethod`. I'm not married to the name though, so suggestions are always welcome. :)

fingolfin commented on gap/base/methods.gi:
```diff
@@ -0,0 +1,59 @@
+
+InstallGlobalFunction(CallRecogMethod,
+function(m, args)
+    return CallFuncList(FunctionOfRecogMethod(m), args);
```
If you change this to do something like `if IsRecord(m) then ... elif IsRecogMethod(m) then ... else Error() fi;` then you could immediately convert code to use it, right?

Alternatively, this could become an operation with several methods. Doesn't really matter, it's only temporary.

fingolfin commented on gap/base/methsel.gi:
```diff
@@ -20,39 +20,46 @@
+            comment := method.comment;
+        fi;
+        method := RecogMethod(method.stamp, comment, method.method);
+        SetComment(method, comment);
```
This is redundant, isn't it?
```suggestion
```

fingolfin commented on gap/base/methsel.gi:
```diff
@@ -81,33 +88,33 @@ InstallGlobalFunction( "CallMethods", function(db, tolerancelimit, methargs...)
+                ms.failedMethods.(db[i].method.stamp) > tolerance then
                 Info(InfoMethSel, 4, "Skipping rank ", db[i].rank,
-                     " method \"", db[i].stamp, "\".");
+                     " method \"", db[i].method.stamp, "\".");
```
If you add methods for `Stamp` and `Comment` using `InstallOtherMethod` which take an `IsRecord`, then perhaps you could immediately change this to
```suggestion
                     " method \"", Stamp(db[i].method), "\".");
```

fingolfin reviewed (commented):
Direction seems fine! Some comments on minor details, some are just opinions.

Don't forget to also adjust the HOWTO once everything is settled :-).

fingolfin commented on gap/base/recognition.gi:
```diff
@@ -451,7 +451,20 @@ InstallGlobalFunction( RecogniseGeneric,
+                    rec(rank := hint.rank,
+                        method := rec(method := hint.method, stamp := hint.stamp)));
+            fi;
+            allmethods := Concatenation(tmp, allmethods);
```
Huh, aren't you adding methods several times this way? You keep adding to tmp, then prepend it to allmethods -- again and again.

So either move this `Concatenation` out of the loop, or else just get rid of `tmp` and replace `Add(tmp` above by `Add(allmethods` (twice) -- the order should not matter as we sort by rank below anyway. 

Well, it might matter if several methods have identical rank, *and* if we were using `StableSortBy`.... I think having methods with equal rank is an error here, though, and perhaps we should add a corresponding assertion at the end. Something like this:
```gap
Assert(0, Length(Set(allmethods, m->m.rank)) = Length(allmethods)); # verify no rank occurs more than once
```

fingolfin replied:
added a fix

fingolfin reviewed (commented):
There is an error were something somewhere access `r.rank` which is not bound...

fingolfin commented on gap/base/methods.gi:
```diff
@@ -0,0 +1,84 @@
+end);
+
+InstallGlobalFunction(RecogMethod,
+function(stamp, comment, func)
```
So there is no support for `validatesOrAlwaysValidInput` yet, which affects conversion of methods which set it.

That said, there isn't a *real* need to set `validatesOrAlwaysValidInput`, is there? It's not as if we are using it for anything. It's just a marker saying "someone went over this method and believes they have established that it is correctly validating everything". The one reason to have it as an explicitly variable is that it helps us to catch methods which don't have it yet?

ssiccha replied:
At the moment, we can use it to check which methods don't have it yet, yes. I think in the long run it should be set to `true` for every method so we can remove it. Should we rather turn it into a source comment?

fingolfin commented on gap/projective/d247.gi:
```diff
@@ -349,7 +349,9 @@ RECOG.SortOutReducibleSecondNormalSubgroup :=
 #! TODO
 #! @EndChunk
-FindHomMethodsProjective.D247 := function(ri,G)
+FindHomMethodsProjective.D247 := RecogMethod("D247",
```
I suggest we add a helper `BindRecogMethod` to be used like this:
```suggestion
BindRecogMethod(FindHomMethodsProjective, "D247",
```

fingolfin replied:
I've now added `BindRecogMethod`

ssiccha commented:
Some commits are independent of this PR. I'll put them into a new PR.

fingolfin commented:
@ssiccha I merged the independent commit (see PR #269) after a few extra tweaks, rebased this PR and cleaned it up.

I think now only `projective.gi` needs to be converted, then at least all calls to `AddMethods` only use RecogMethod objects. After these are done, AddMethods can be simplified again. Then there are likely in addition a few more methods only used in hints; once those are converted, the `allmethods` code can be simplified again, plus a bunch of other compatibility code be removed.

The `\=` method you added yesterday as a workaround for making `regen_doc.g` work can then also be removed again, and `regen_doc.g` be changed to instead just compare the stamp of methods to decide whether they are "equal".

fingolfin commented on gap/generic/KnownNilpotent.gi:
```diff
@@ -104,14 +106,12 @@ FindHomMethodsGeneric.KnownNilpotent := function(ri,G)
   Add(forkernel(ri).hints,
-      rec( method := FindHomMethodsGeneric.KnownNilpotent, rank := 4000,
-           stamp := "KnownNilpotent" ));
+      rec( method := FindHomMethodsGeneric.KnownNilpotent, rank := 4000 ) );
```
I wonder if we now also use `AddMethod` for these places? Like this:
```gap
      AddMethod(forkernel(ri).hints, FindHomMethodsGeneric.KnownNilpotent, 4000);
```

fingolfin commented on gap/base/methsel.gi:
```diff
@@ -20,39 +20,45 @@
     fi;
     l := Length(methodDb);
-    pos := First([1 .. l], i -> methodDb[i].rank <= method.rank);
+    pos := First([1 .. l], i -> methodDb[i].rank <= rank);
     if pos = fail then
         pos := l + 1;
     fi;
```
I think this also works (but didn't test, beware of off-by-one errors etc.)
```suggestion
    pos := PositionSortedBy(methodDb, -rank, m -> -m.rank);
```

fingolfin commented on gap/base/methsel.gi:
```diff
@@ -20,39 +20,45 @@
     fi;
     l := Length(methodDb);
-    pos := First([1 .. l], i -> methodDb[i].rank <= method.rank);
+    pos := First([1 .. l], i -> methodDb[i].rank <= rank);
     if pos = fail then
         pos := l + 1;
     fi;
```
I think this also works (but didn't test, beware of off-by-one errors etc.)
```suggestion
    pos := PositionSortedBy(methodDb, -rank, m -> -m.rank);
```
and then the local variable `l` could be removed.

ssiccha replied:
I'm going to install a method for `\<` for `RecogMethod`s, unless you object. Then we can use `PositionSorted` etc.

fingolfin replied:
But why? We are comparing ranks, not recognition methods

ssiccha commented:
I think I'm finished with everything except for adjusting the method how to and double-checking whether I forgot to remove any hacks.

fingolfin commented:
@ssiccha regarding the commit "Make AddMethod compatible with GAP 4.10" -- Is there any value in keeping compatibility with GAP 4.10 at this point? We may want to use other improvements from 4.11 (or even 4.12, once it is out) anyway.

ssiccha commented:
@fingolfin I'd like to document `RecogMethod` using AutoDoc. However, with what I did in the last commit, the documentation of `RecogMethod` is put into a new chapter `methsel`. Do you know of a way to tell AutoDoc to include it into the chapter with GapDoc label `methsel`? Or is just not intended to start a chapter via GapDoc and then put stuff there via AutoDoc? I'm super confused. ^^

ssiccha commented:
I'm just going to use GAPDoc. I have no idea how to include it with AutoDoc doc.

fingolfin commented:
@ssicha Will you complete this? When?

Regarding AutoDoc and manuals: just write `@Chapter TITLE-OF-CHAPTER` before / at the start of your AutoDoc comment and you are good to go. No, there is no way to refer to a chapter by label, you need to repeat its title, but that's usually the same anyway.

ssiccha commented:
I've had a big teaching backlog to work through, but I'm done now.

I think I'm a bit dense when it comes to how to make GapDoc and AutoDoc work together. I've tried all possible combinations for `TITLE-OF-CHAPTER` I could think of:
`Method selection`, `Method_selection`, `Chapter_Method_selection`, ... The global function `RecogMethod` is always put into a completely new chapter. What am I doing wrong? For now I'm just going to use `@BeginChunk`.

ssiccha commented:
I've made some progress on the documentation. I'll improve it a little bit further, then this should be good.

fingolfin commented:
Sorry, I misunderstood: there is no way to write stuff in AutoDoc and "magically" insert it into an existing GAPDoc chapter. The only way to do that is indeed to use chunks. What I was talking about is only for adding content to a chapter "managed" by AutoDoc (which I should have realized is not the case here, but didn't).

ssiccha commented:
Rebased onto master. This is now ready for review @fingolfin.

FriedrichRober commented on gap/base/methods.gd:
```diff
@@ -0,0 +1,130 @@
+# <ManSection>
+# <Func Name="RecogMethod" Arg="stamp, comment, func"/>
+# <Description>
+# Return a recognition method <C>method</C> in the filter
```
Maybe add a comment here, that `RecogMethod` should not be called directly but rather be created via `BindRecogMethod`.

ssiccha commented:
I've added a few tests just to get the coverage of `gap/base/methods.gi` to 100%.

FriedrichRober commented on gap/base/methods.gd:
```diff
@@ -0,0 +1,130 @@
+# <ManSection>
+# <Func Name="CallRecogMethod" Arg="m, args"/>
+# <Description>
+# Call <C>UnpackRecogMethod(m)</C> with arguments <A>args</A> and return the
```
Maybe add a comment here that `m(args)` does essentially the same.

FriedrichRober replied:
After reading the review of Max, probably it would be better to drop `m(args)` entirely.

ssiccha replied:
I've removed the `m(args)` functionality as @fingolfin suggested.

fingolfin commented on gap/base/methods.gd:
```diff
@@ -0,0 +1,130 @@
+# </Description>
+# </ManSection>
+# <#/GAPDoc>
+DeclareGlobalFunction("UnpackRecogMethod");
```
Is this really something we should expose to the user? What for? Once it's out, it's harder to take it back.

ssiccha replied:
I've removed the documentation and the `Declare` statement.

FriedrichRober commented on doc/methsel.xml:
```diff
@@ -100,6 +103,31 @@ function:
+<#Include Label="FindHomMethodsProjective">
+<#Include Label="FindHomMethodsGeneric">
+
+This global record holds the functions for writing group
```
This line is confusing. To what global record are you referring? And why only matrix group recognition.

ssiccha replied:
Ooh, yes. Thanks for spotting that.

fingolfin commented on gap/base/methods.gi:
```diff
@@ -0,0 +1,90 @@
+##
+#############################################################################
+
+InstallMethod(CallFuncList, [IsRecogMethod, IsList],
+{f, args} -> CallFuncList(f!.func, args));
+
```
Do we really want this? I think you are using `CallRecogMethod` explicitly after all, and I think this is indeed better -- it is easy to grep for `CallRecogMethod` and it is then possible to track what happens. But if one replace `CallRecogMethod(meth, val)` by `meth(val)` then this is rather intransparent, and may be very confusing, esp. to people who don't know that one can overload function call syntax in GAP (i.e., >99.9% of them)
```suggestion
```

FriedrichRober replied:
I agree, we should probably drop `meth(val)`.

ssiccha replied:
Makes sense. I've removed this.

fingolfin commented on gap/base/methsel.gi:
```diff
@@ -20,40 +20,21 @@
-    if not IsBound(method.comment) then
-        method.comment := "";
+    # This may be turned into an error in the future, see github issue #184.
+    if not HasValidatesOrAlwaysValidInput(method) then
```
I just merged another PR of your where you removed this code, I think?

ssiccha replied:
I rebased it. Now it's gone here too.

fingolfin commented on gap/generic/KnownNilpotent.gi:
```diff
@@ -76,7 +76,9 @@ end;
 #! @EndChunk
-FindHomMethodsGeneric.KnownNilpotent := function(ri,G)
+BindRecogMethod(FindHomMethodsGeneric, "KnownNilpotent",
+"TODO",
```
Perhaps
```suggestion
"method for nilpotent groups which are not p-groups",
```

fingolfin commented on gap/matrix.gi:
```diff
@@ -125,7 +126,9 @@ end;
 #! @EndChunk
-FindHomMethodsMatrix.Scalar := function(ri, G)
+BindRecogMethod(FindHomMethodsMatrix, "Scalar",
+"Hint TODO",
```
Not for this PR, but: we should document these methods

fingolfin commented on gap/matrix.gi:
```diff
@@ -214,7 +217,9 @@ end;
 #! @EndChunk
-FindHomMethodsMatrix.BlockScalar := function(ri, G)
+BindRecogMethod(FindHomMethodsMatrix, "BlockScalar",
+"Hint TODO",
```
Not for this PR, but: we should document these methods

fingolfin commented on gap/matrix.gi:
```diff
@@ -479,7 +478,9 @@ end;
 #! @EndChunk
-FindHomMethodsMatrix.BlockLowerTriangular := function(ri,G)
+BindRecogMethod(FindHomMethodsMatrix, "BlockLowerTriangular",
+"Hint TODO",
```
```suggestion
"for a group generated by block lower triangular matrices",
```

fingolfin commented on gap/matrix.gi:
```diff
@@ -525,18 +523,17 @@ end;
 #! @EndChunk
-FindHomMethodsMatrix.BlockDiagonal := function(ri,G)
+BindRecogMethod(FindHomMethodsMatrix, "BlockDiagonal",
+"Hint TODO",
```
```suggestion
"for groups generated by block diagonal matrices",
```

fingolfin commented:
@ssiccha hmmm, now all tests are failing

ssiccha commented:
Jepp. I'm looking into it. I suspect its because I rebased everything on top of master.

ssiccha commented:
I added, but didn't document, the ability to call a recog method via `m(args)`. It's used in several places of recog. I also adjusted the SnAnUnknownDegree code to this PR. Tests pass on my machine now. Let's see whether they also pass on GitHub. Then I'll merge this.
